zip-0308.rst

```diff
@@ -0,0 +1,326 @@
+----------------------
+
+We assume use of the normal wallet note selection algorithm and change
+handling. Change is sent back to the default address, which is the z-address
+of the first selected Sprout note. The number of JoinSplits will therefore be
+the same as for a normal transaction sending the same amount with the same
+wallet state. Only the ``vpub_out`` of the last JoinSplit will be nonzero.
+There will always be exactly one Sapling Output.
+
+The expiration height is chosen as for other transactions.
+
+The fee for each migration transaction MUST be 0.0001 ZEC.
+
+Some wallets by default add a "developer fee" to each transaction, directed
+to the developer(s) of the wallet. This is typically implemented by adding
+the developer address as an explicit output, so if migration transactions are
+generated internally by ``zcashd``, they will not include the developer fee.
+We strongly recommend *not* patching the ``zcashd`` code to add the developer
+fee output to migration transactions, because doing so partitions the anonymity
+set between users of that wallet and other users.
+
+There MUST NOT be any transparent inputs or outputs, or Sapling Spends, in a
+migration transaction.
+
+The ``lock_time`` field MUST be set to 0 (unused).
+
+
+Open questions
+--------------
+
+The above strategy has several "magic number" parameters:
+
+* the interval between batches (500 blocks)
+* the maximum number of transactions in a batch (5)
+* the distribution of exponents (uniform integer in 6..9)
+* the distribution of mantissae (uniform integer in 1..99).
+
+These have been chosen by guesswork. Should we change any of them?
+
+In particular, if the amount to migrate is large, then this strategy can
+result in large amounts (up to 990 ZEC, worth USD ~49500 at time of writing)
+transferred in each transaction. This leaks the fact that the transaction was
+sent by a user who has at least that amount.
+
+The strategy does not migrate any remaining fractional amount less than
+0.01 ZEC (worth USD ~0.50 at time of writing). Is this reasonable?
+
+In deciding the amount to send in each transaction, the strategy does not
+take  account of the values of individual Sprout notes, only the total amount
+remaining to migrate. Can a strategy that is sensitive to individual note
+values improve privacy?
+
+An adversary may attempt to interfere with the view of the block chain seen
+by a subset of nodes that are performing migrations, in order to cause those
+nodes to send migration batches at a different time, so that they may be
+distinguished. Is there anything further we can do to mitigate this
+vulnerability?
+
+
+RPC calls
+---------
+
+Nodes MUST maintain a boolean state variable recording whether migration is
+enabled. This can be set or reset using the following RPC call::
+
+  setenablemigration true|false
+
+It is intentional that there is no facility to change options associated with
+the migration, since such options can distinguish users.
```

Comment:
This leaves it unclear as to _where_ the migrated funds end up. I agree that minimizing the configuration is a good idea, but would it be useful to allow any Sapling address to be specified as the target? Or possibly any Sapling address with the proviso that it is in the local wallet? Or do we just say "migrated funds will end up in account 0 of the wallet"? In this case, we'd need to automatically generate the first Sapling address if the user had not done so already, but that seems reasonable.

Reply:
There's no privacy reason not to do this, but:
a) it is simpler to have no configuration
b) the wallet needs to be able to tell which transactions are migration transactions in order to implement the getmigrationstatus RPC.

But as long as there is still a way to do b), then I'm fine with allowing it.

Reply:
Fixed by adding `-migrationdestaddress=<zaddr>`.